Update Gutenberg packages after v12.5.0 release

westonruter opened the pull request:
- [x] Verify Preview button shows up in older versions of WordPress without Gutenberg as well.

pierlon committed:
Update Gutenberg package dependencies

westonruter committed:
Fix eslint errors

pierlon committed:
Attempt to fix error to resolve a module for a custom reporter

westonruter committed:
Update snapshots

pierlon committed:
Merge branch 'develop' into update/gutenberg-v12.4.1-packages

pierlon committed:
Update Gutenberg package dependencies

pierlon committed:
Merge branch 'develop' into update/gutenberg-v12.4.1-packages

pierlon committed:
Update Gutenberg package dependencies

westonruter committed:
Fix JS import order

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into update/gutenberg-v12.4.1-packages

* 'develop' of github.com:ampproject/amp-wp:
  Bump node-fetch from 2.6.1 to 2.6.7 in /.github/actions/draft-release
  Bump lint-staged from 11.2.6 to 12.3.3
  Bump mustache/mustache from 2.13.0 to 2.14.1
  Bump cweagans/composer-patches from 1.7.1 to 1.7.2
  Bump follow-redirects from 1.14.5 to 1.14.7
  Bump webpackbar from 5.0.0-3 to 5.0.2

pierlon committed:
Update Gutenberg package dependencies

westonruter committed:
Merge branch 'update/gutenberg-v12.4.1-packages' into update/gutenberg-v12.5.0-packages

* update/gutenberg-v12.4.1-packages:
  Fix JS import order
  Update Gutenberg package dependencies
  Update Gutenberg package dependencies
  Update snapshots
  Attempt to fix error to resolve a module for a custom reporter
  Fix eslint errors
  Update Gutenberg package dependencies

westonruter commented:
https://github.com/ampproject/amp-wp/pull/6858#issuecomment-1028378641:

> The E2E test is failing due to the `.editor-post-preview` element being _after_ the `.amp-wrapper-post-preview` element, whereas previously it was _before_ (and it is being inserted as such). I saw this locally as well for a bit, but now I don't any longer 😕

https://github.com/ampproject/amp-wp/pull/6858#issuecomment-1028457811:

> In testing this with Gutenberg 12.5, it seems the problem is that `WrappedAmpPreviewButton` is constructed _before_ the `.editor-post-preview` element has been added to the DOM, so `this.postPreviewButton` is `null` here:
>
> https://github.com/ampproject/amp-wp/blob/4e28debd74160547d896881dec64980fb123363d/assets/src/block-editor/plugins/wrapped-amp-preview-button.js#L34

https://github.com/ampproject/amp-wp/runs/5046692459?check_suite_focus=true#step:12:1

![image](https://user-images.githubusercontent.com/134745/152279033-1211b0e2-d306-47b3-b08e-d90efe03566b.png)

delawski committed:
Rewrite AMP Preview Button React class into functional component

delawski commented:
> In testing this with Gutenberg 12.5, it seems the problem is that `WrappedAmpPreviewButton` is constructed _before_ the `.editor-post-preview` element has been added to the DOM

That's interesting. Mu gut feeling was that it has something to do with the fact that we're using class components for the preview button. It seems like they are now being constructed before the editor is set up.

So I went ahead and rewritten the class component into a functional component. It seems that it helped. The preview button is now back in its original position.

However, there's a slight regression when Gutenberg is turned off. The preview button is added a bit too late causing a horizontal layout shift:

https://user-images.githubusercontent.com/478735/152365413-8d69584b-e59f-4690-b88e-b952bfcc3f2d.mov

I'm going to dive deeper and see if this can be easily fixed.

westonruter commented on assets/src/block-editor/plugins/wrapped-amp-preview-button.js:
```diff
@@ -19,74 +13,43 @@ import AmpPreviewButton from '../components/amp-preview-button';
-	}
+		return () => {
+			if ( postPreviewButton.current && root.current ) {
+				postPreviewButton.current.parentNode.removeChild( root.current );
```
Should this also set `postPreviewButton.current` and `root.current` to null?

westonruter replied:
Or rather, just `root.current`?

westonruter replied:
Or yeah, both right? As otherwise if it gets re-rendered then the logic above will never run since `! root.current && ! postPreviewButton.current`.

delawski replied:
Yes, they should be nulled here.

delawski replied:
This has been addressed in 9f3e1c8.

delawski committed:
Use Publish button as anchor for AMP Preview button

The post Publish/Update button is rendered early on (contrary to post Preview button) and so it's safer to use it to imperatively add an AMP Preview button to the block editor markup.

Also, the `viewable` property of a post type get defined late, similarly to when the post Preview button gets added. That's why we assume a post type is viewable and show the AMP Preview button initially. Only if we learn that a post type is not viewable, we hide the AMP Preview button.

delawski commented:
@westonruter I think I found a way to prevent the flicker, at least for the majority of use cases. Check out a message  in 9f3e1c8.

delawski commented:
Oh, there's a side-effect of the recent changes that also makes the E2E tests fail. The AMP preview button is now in a different location on mobile:

![Screenshot 2022-02-03 at 19 00 18](https://user-images.githubusercontent.com/478735/152401575-086ec01b-a10a-4e4b-a69e-ba38a0559291.png)

I'll quickly check out if there's a way to put it back in place.

westonruter commented:
In terms of the shift, this is looking good with Gutenberg and without Gutenberg, and when the `post` post type is enabled and disabled. The following videos all show the same good behavior:

https://user-images.githubusercontent.com/134745/152401770-7990c333-89c3-4dc0-ba69-46a22c45ca18.mov

https://user-images.githubusercontent.com/134745/152401775-7e703ea8-952b-4630-89b0-6e8d09d24589.mov

https://user-images.githubusercontent.com/134745/152401779-be1c2b6c-1634-4725-b935-41bdfa9c6800.mov

https://user-images.githubusercontent.com/134745/152401784-64051158-0912-4cc2-b56b-d2d80daf4557.mov

In the case for when the post type is disabled, the scripts aren't even enqueued:

https://github.com/ampproject/amp-wp/blob/36edf69661e8ee6c7c35bce1c483c9524dbf386f/includes/admin/class-amp-post-meta-box.php#L226-L230

westonruter commented:
And the only post types that are eligible for being AMP-enabled are those which have `'public' => true`:

https://github.com/ampproject/amp-wp/blob/36edf69661e8ee6c7c35bce1c483c9524dbf386f/includes/class-amp-post-type-support.php#L44-L64

It's true that a post type could have `public` set to true but have `publicly_queryable` set to false, I think, and in that case it may be possible that we could accidentally show or hide the preview button. It would be marginally better to here instead do the following since there is now a `is_post_type_viewable` filter in WP 5.9:

```diff
diff --git a/includes/class-amp-post-type-support.php b/includes/class-amp-post-type-support.php
index 3d4560641..17eb6519b 100644
--- a/includes/class-amp-post-type-support.php
+++ b/includes/class-amp-post-type-support.php
@@ -42,14 +42,9 @@ public static function get_builtin_supported_post_types() {
 	 * @return string[] Post types eligible for AMP.
 	 */
 	public static function get_eligible_post_types() {
-		$post_types = array_values(
-			get_post_types(
-				[
-					'public' => true,
-				],
-				'names'
-			)
-		);
+		$post_types = array_values( get_post_types( [], 'names' ) );
+
+		$post_types = array_filter( $post_types, 'is_post_type_viewable' );
 
 		/**
 		 * Filters the list of post types which may be supported for AMP.
```

But that can be done later if needed. I'll open an issue.

delawski committed:
Reposition AMP Preview button once block editor is fully loaded

delawski commented:
I've pushed another update (15c713c). This time, we're adding the AMP Preview button in two rounds.

As long as the Preview button (the one that we've used so far) is not in the DOM, we'll adding our AMP Preview button before Publish/Update button.

Right after the Preview button is added to the editor (we use `isViewable` as an imperfect indicator of that), the AMP Preview button gets re-rendered in the correct location.

I haven't ran E2E tests locally. It may turn out that E2E tests are "too quick" and will find our AMP Preview button before it gets mounted in the correct location. In such a case, we can increase the `wait` time for the particular assessment or make the AMP Preview button selector more _liberal_ (e.g. just check if the button is a child of `.edit-post-header__settings` element, and not try confirming its exact location).

westonruter commented:
> make the AMP Preview button selector more _liberal_ (e.g. just check if the button is a child of `.edit-post-header__settings` element, and not try confirming its exact location).

I was also considering this myself.

westonruter commented:
It passed!

westonruter commented:
@delawski I've just noticed an issue: the preview button is not showing up in WP 5.6, 5.7, and 5.8 until the user makes a change and then saves a draft. 

https://user-images.githubusercontent.com/134745/152435922-22360a24-892e-441e-8577-f17f71af929d.mov

It seems the container is empty:

![image](https://user-images.githubusercontent.com/134745/152436149-37b3741c-117e-42d5-bc53-ee948773316c.png)

And also, when I load a post that is _published_ I see the button show up for a moment and then disappear until I make a change and save the post:

https://user-images.githubusercontent.com/134745/152436429-7a6ba4fd-e53b-4017-8dc5-91a896a2e1bb.mov

It doesn't happen in either 5.9 or in 5.7 or 5.8 with the latest Gutenberg is active.

So I don't consider this a blocker for 2.2.1, but it should be fixed in 2.2.2.
